### ISSUE TITLE
Placeholder templater should optionally supply default values for missing parameters

### ISSUE THREAD

**Issue by barrywhart** [sqlfluff/sqlfluff#2108]
Capturing some thoughts from a PR (#2107) that we decided was too specific an implementation, but could potentially be re-implemented in a more generic way.

I've fixed some issues as pointed out by @jpy-git, and added some basic support for Postgres data types.

The author of the PR, @danicc097, described his use case:

>I'm using [sqlc](https://github.com/kyleconroy/sqlc) with Postgres. Variables use `@` prefix, so I'd have queries like:

```sql
-- name: UpdateUserById :one
UPDATE
  "users"
SET
  "password" = CASE WHEN @password_do_update::boolean THEN
    @password
  ELSE
    "password"
  END,
  "salt" = CASE WHEN @salt_do_update::boolean THEN
    @salt
  ELSE
    "salt"
  END,
  "username" = CASE WHEN @username_do_update::boolean THEN
    @username
  ELSE
    "username"
  END,
  "email" = CASE WHEN @email_do_update::boolean THEN
    @email
  ELSE
    "email"
  END
WHERE
  "id" = @id
RETURNING
  *;
```

>Creating dummy values in the config seems unneccesary in these cases, except common things like "id", etc.  Regarding usefulness, if a user opts in to autofill parameters it will look for the Postgres "::" syntax (else notify it will be replaced with text) and parse simple builtin types and arrays. Its just a matter of extending the recognized types with dummy values in the templater. Of course, as it is now, the implementation is limited to Postgres (else it will always autofill a string); I don't know if for other databases it's worth both the verbosity of typing and effort to make its type parser in the templater, maybe someone has more experience in other dialects.

==

Here's one high-level thought on how we might design this generically:
Let the user define (in `.sqlfluff`) a set of prefixes or suffixes that implicitly specify the type and default value of the variable, e.g. (this is adapted from the original PR):
```
[sqlfluff:templater:placeholder:context:autofill:prefixes]
str="'string'"
date=CURRENT_DATE
tbl=table_$1
col=column_$2
```

In other words, a missing variable with a "str" prefix gets a quoted string as a default value,  "date" prefix gets a date literal, and "tbl" and "col" prefixes get a not-quoted name followed by the variable name (to avoid having every table or column name be the same, which could confuse the linting rules.)

==

Another idea (more complex):
If we wanted to support more complex cases of default behavior, we could allow calling user-provided Python code to influence the mapping of undefined variable names to config values or let them return default values directly. This would be similar to the way the Jinja templater lets users provide modules whose functions are mapped into the Jinja context, making them available to macros.

This is an interesting idea, but more of a "power user" approach. While the Jinja templater primarily exposes Jinja constructs, here we'd be extending a very simple templater with concepts we invented ourselves, which means we'd take on more of the documentation and maintenance burden. Overall, the first idea seems more appealing than the second, in terms of the right amount of power and complexity.

**Comment by jpy-git** [comment]
@barrywhart I think the first idea is the most appropriate as its pretty simple for the user to setup and configure + doesn't require them to change anything about the way they write their sql. I think if they had this enabled and a variable wasn't given one of the specified prefixes then it should still raise the variable not found error rather than trying any fancy substitution assumption to avoid ambiguity.

My issue with option 2 would be: The point of a feature like this would be for those who don't really care too much about the contents of  the substitution, just that it has the correct type. My question would be, is introducing this sort of complexity any easier than just taking the 10 secs to put a value in `.sqlfluff`?
